### ISSUE TITLE
Questions on epistemic uncertainty

### ISSUE THREAD

**Issue by TomLucidor** [strangehospital/Frontier-Dynamics-Project#1]
Saw your post here https://www.reddit.com/r/ControlProblem/comments/1r88r5d/can_we_model_ai_epistemic_uncertainty/

I got a few questions on the idea of epistemic humility:
1. If the dimensional space is too small, would these ML system still be able to draw a line between IID and OOD?
2. Can ML detect the shape of data by examining the distribution of data relative to absolutely random distributions?
3. How can this type of ML system handle embedding/LLM/Diffusion with extreme quantity of dimensions?
4. Does architecture matter for effectiveness of technique e.g. CNN and SWA, RNN and Linear Attention, deep MLP layers?

**Comment by strangehospital** [comment]
Hello, thank you for the questions. Keep posted for updates on STLE.v3!

# 1) Yes, the ML system will still be able to draw a line, but with reduced precision. The quality degrades gracefully, but not catastrophically. There is a trade-off that must be made between high and low dimensionality. For example, low dimensionality comes with the benefits of faster computation, but at the cost of significant overlap, making IID-OOD boundaries fuzzy. For high dimensionality, you have more "room" to separate concepts more cleanly, but more likely to face the curse of dimensionality

Latest Research as of (Feb 22th): 
--> Key insight: STLE doesn't require a specific dimensionality. It adapts to the intrinsic dimensionality of your data.

Our findings (from the Sky Project "MarvinBot/STLE.v3"):

Raw embeddings: 384 dimensions (from sentence-transformers)
Intrinsic dimensionality: ~165 dimensions (90% of variance)
Projection used: 64 dimensions (learned, not fixed)

-8-D space example
-Training: 100 "Chemistry" samples cluster in region A
-Query: "Organic chemistry" falls in region A → IID (μ_x ≈ 0.85)
-Query: "Ancient Rome" falls far from region A → OOD (μ_x ≈ 0.15)

Results: Even in 8-D, STLE can distinguish IID vs. OOD. The boundary is less sharp than in 64-D, but it's still meaningful.

Empirical Evidence:
Experiment (from Posterior Networks paper, Charpentier et al., 2020):

MNIST (10 digit classes) in 2-D latent space: AUROC = 0.82 for OOD detection
MNIST in 32-D latent space: AUROC = 0.94 for OOD detection
Interpretation: Higher dimensions help, but even 2-D provides useful IID/OOD separation.

Therefore, one can argue that STLE's advantage is that the μ_x (accessibility) score gives you a continuous confidence measure, not just a binary yes/no:
μ_x = 0.95 → "I'm very confident this is IID"
μ_x = 0.60 → "This is on the boundary (frontier)"
μ_x = 0.10 → "This looks OOD to me"


# 2 Yes, It's the core of STLE's density estimation and exactly what normalizing flows do. Here's how it works:

STLE will: 
Step 1) Learn the "shape" of your training data (via normalizing flows)
Step 2) Compare new queries to this shape
Step 3) If a query fits the shape → IID (high μ_x)
Step 4) If a query doesn't fit → OOD (low μ_x)

Normalizing flows transform a simple distribution (Gaussian) into a complex one (your data):

Simple Gaussian → Flow transformations → Complex data distribution
     (sphere)              (learned)              (banana shape)

Comparison to Random Distributions:
STLE explicitly models the "inaccessible" (OOD) space as having low density or uniform distribution:

μ_x(r) = (N_x · P(r | learned_shape)) / 
         (N_x · P(r | learned_shape) + N_y · P(r | random_uniform))


# 3) To handle embedding/LLM/Diffusion with extreme quantity of dimensions, use learned dimensionality reduction + evidence scaling. We recommended this because this is exactly what we implemented in our ML learning training with success. 

TLDR:
Handling Extreme Dimensions (i.e 12,000-D LLM embeddings)

Strategy:
--> Use pre-trained embeddings (freeze the LLM encoder)
--> Add trainable projection (12,000-D → 64-D or 128-D)
--> Train projection + flows jointly with PAC-Bayes objective
--> Use evidence scaling to prevent saturation

Modern embeddings:

-LLMs (e.g., GPT-4): 12,000+ dimensions
-Sentence-Transformers: 384-768 dimensions
-Diffusion models (e.g., Stable Diffusion): 1024+ dimensions

Problem: Normalizing flows struggle in such high dimensions because:

-Data sparsity: Training data becomes sparse (curse of dimensionality)
-Computational cost: O(d²) or O(d³) for flow layers (d = dimensions)
-Density estimation fails: All densities → 0 (numerical underflow)

Solution: Two-Stage Approach (what worked for us)

Stage 1: Learned Projection (384-D --> 64-D)
Trainable neural network layer
projection = nn.Linear(384, 64)

-Learns to compress 384-D embeddings into 64-D space
-while preserving relevant structure for STLE

Stage 2:  Normalizing flows in 64D
Train flows on the compressed space
flow = NormalizingFlow(latent_dim=64, num_layers=4)

The key technique is Evidence Scaling (λ). Please note, this required an update to STLE's Bayesian formula. Visit my post on r/LLMphysics for an early preview of formula: https://www.reddit.com/r/LLMPhysics/comments/1rbns7p/the_redemption_of_crank_a_framework_bros/

Problem: Even in 64-D, with 16,000+ training samples, densities can saturate (all μ_x → 1.0).

Solution: Scale the evidence contribution
-Instead of:
α = β + N_x · p(z)  # Saturates when N_x is large

-Use:
α = β + λ · N_x · p(z)  # λ ≈ 0.001 prevents saturation

Effect: Controls how "confident" the system becomes as it sees more data.
--> Without λ: Overconfident (all μ_x → 1.0)
--> With λ: Calibrated (μ_x ranges from 0.1 to 0.9)

Handling Extreme Dimensions (e.g., 12,000-D LLM embeddings)

Example Architecture:

Text Input
    ↓
[Frozen LLM Encoder] → 12,000-D embedding
    ↓
[Trainable Projection] → 64-D latent
    ↓
[Normalizing Flows] → Density p(z)
    ↓
[STLE Formula] → μ_x accessibility


# 4) STLE is essentially a meta-framework and is "architecture-agnostic," but some architectures are probably better suited for specific data types. It doesn't care about the underlying architecture because it operates on learned representations (i.e embeddings).

General structure:
Raw Data → [Architecture] → Embedding → [STLE] → μ_x

ex)
Data Type: Images -> Best Architecture: CNN -> Why: it captures spatial structure -> STLE on CNN feature maps (ex ResNet-18 output)

Data Type: Text -> Best Architecture: Transformer -> Why: captures long range dependencies -> STLE on BERT/GPT embeddings 

Example: CNN vs. MLP for MNIST
CNN-based STLE:
-Architecture
cnn_encoder = nn.Sequential(
    nn.Conv2d(1, 32, 3, padding=1),
    nn.ReLU(),
    nn.MaxPool2d(2),
    nn.Conv2d(32, 64, 3, padding=1),
    nn.ReLU(),
    nn.MaxPool2d(2),
    nn.Flatten(),
    nn.Linear(64*7*7, 64)  # Latent space
)

Apply STLE on CNN embeddings
latent = cnn_encoder(image)
mu_x = stle.estimate_accessibility(latent)

Architecture Selection Guidelines for STLE

# Good Architectures for STLE:
-Inductive bias matches data structure (CNN for images, Transformer for text)
-Moderate capacity (not too shallow, not too deep)
-Pre-trained when possible (e.g., BERT, ResNet) → better embeddings

# Architectures to Avoid:
-Too shallow (won't learn meaningful embeddings)
-Too deep without regularization (overfit, poor generalization)
-Mismatched to data type (MLP for images, CNN for text)

Hope that was helpful,

strangehospital.
